lib/db/leveldb_dbinstance.go

```diff
@@ -444,11 +435,15 @@ func (db *Instance) withNeed(folder, device []byte, truncate bool, fn Iterator)
 
 		have := false // If we have the file, any version
 		need := false // If we have a lower version of the file
-		var haveVersion protocol.Vector
+		var haveFileVersion FileVersion
 		for _, v := range vl.Versions {
 			if bytes.Equal(v.Device, device) {
 				have = true
-				haveVersion = v.Version
+				haveFileVersion = v
+				if v.Invalid {
```

Comment:
I'm suspicious. Why does the need decision depend only on the invalidness and not at all on the version comparison etc that happens later?

Reply:
The version isn't relevant for invalid files, e.g. when ignores changed we want to get all invalid files (`needinvalid == true`) while on "normal" pulls we don't consider invalid files as needed. The latter is the same behaviour as before this PR where invalid files weren't in the global version list, i.e. `have == false` which is equivalent to `need == true`.

Reply:
But previously we would just slide over the invalid files, continuing along the list of available versions. With this, I guess we might have a version list of `[{version=42, invalid=true}, {version=42, invalid=false}]` and when running with `needInvalid=false` we would then immediately conclude we don't need the file, while in the old code we would see the second entry and actually evaluate the version to see if we need it or not.

Reply:
We first check whether the version is from the device in question, so there is only one version to be considered. Also before this we did break out of the loop regardless of versions. Now we do find a version for this device even if it is invalid, so we need to decide whether to include (new behaviour) or ignore (old behaviour) this invalid file.

Reply:
While I still think this would work if invalid was equal to ignored, there is a problem with symlinks also being marked as invalid. I will have to rethink this.